# Fermentation-bacteria

The scripts used as part of this project are here described in order:

## initialiser.py
**Local**: *Used after downloading whole-genomes sequences from NCBI.*

- Sets up the working directory.
- Fills a **Pandas** dataframe with information on the genomes using the "data_summary.tsv" file present in the folder downloaded from **NCBI**.
- The dataframe is filled with informations such as the organism qualifier, taxonomic ID, assembly name/accession, size, gene count, BioProject ID, BioSample ID, etc.
- Removes unnecessary files.
- Moves all the genome sequences to the bacterium-specific folder (removes unnecessary nested folders).
- Saves the full bacteria log data as an **Excel** sheet --> "bacteria_log.xlsx"


## [CheckM](https://github.com/Ecogenomics/CheckM)
**Mjolnir**: *Used after exporting the whole-genome sequence files.*

~~~
#!/bin/sh
#SBATCH -c 8 --mem 40G --output=Acetic.xmfa --time 14-0:00
module load checkm-genome/1.1.3
p=/projects/mjolnir1/people/vhp327/new/
cd $p
mkdir output
for i in $(ls $p)
do if [ -d $i ]
    then
        checkm lineage_wf -x fna $i ${i}_output
        cd ${i}_output
        checkm qa lineage.ms ./ -f qa_${i}.txt
        mv qa_${i}.txt ../output
        cd $p
    fi
done
~~~

**What does CheckM do?**
- Assesses the quality of microbial genomes recovered from isolates, single cells, and metagenomes.
- Uses **Prodigal** to identify genes in a given sequence.
- Places genome bins into a reference genome tree and relies on that phylogenetic placement to identify sets of genes that should be in the genome.
- Looks for those genes using **Hmmer** to search the genome. 
- The completeness is an estimate of the fraction of genes that are expected to be there which were actually found. 
- The contamination is based on identifying the number of single-copy genes, that should only be there once.
- Returns quality assessment files include the percentage of completeness and the degree of contamination for each given genome.
- The output files are saved in a directory named "output".


## qa_filtering.py
**Local**: *Used after importing the "output" directory locally.*

- Reads the quality assessment output files generated by **CheckM**.
- Converts all the qa data into a **Pandas** dataframe that is then converted into an **Excel** sheet --> "total_qa_df.xlsx".

- Reads the "total_qa_df.xlsx" sheet.
- Based on the completeness and the percentage of contamination of each genome, given the predetermined threshold parameters (completeness threshold = 0.9; contamination threshold = 0.05), filters out genomes that are not fit for the analysis. 
- This is done by creating a new directory called "Genomes_filtered" where only the valid whole-genome sequences are present.


## acetic_or_lactic.py
**Mjolnir**: *Used after exporting the "Genomes_filtered" directory to the Mjolnir server.*

- Writes a file ("sort_file.txt") that lists all the genomes and indicates whether each genome is part of the Lactic or the Acetic bacteria family.

- Reads the "sort_file.txt" file that assigns every genome to either the Lactic or the Acetic bacteria family.
- Correspondingly moves the genome sequence files in the "Acetic" folder or the "Lactic" folder. 
- That way, lactic and acetic bacteria can be treated separately for later steps of the analysis.


## first_tree.py
**Local**: *Used after separating the acetic bacteria from the lactic bacteria and filtering out contaminated and incomplete genomes.*

- Retrieves the percentage of completeness of all genomes from the "bacteria log v2.xlsx" Excel sheet (the updated version of "bacteria_log.xlsx").
- For each species of interest (acetic or lactic), finds the file corresponding to the most complete genome of that species.
- Creates a new directory ("Acetic_unique"/"Lactic_unique") containing only the most complete genome per species.

This directory is then exported to the server so that we can proceed to the following stages with **Anvi'o**.

## Pangenome tree with Anvi'o

## Single-copy core gene

